examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+            self.char_to_idx[ch] = i
+
+    def __call__(self, text):
+        text = text.lower()
+        text = text[: self.max_len - 2]
+        text = "<" + text + ">"
+        pad_len = self.max_len - len(text)
+        return [self.char_to_idx.get(ch, 1) for ch in text] + [0] * pad_len
+
+    def get_vocabulary(self):
+        return self.vocab
+
+
+max_target_len = 200  # all transcripts in out data are < 200 characters
+data = get_data(wavs, id_to_text, max_target_len)
+vectorize_layer = VectorizeChar(max_target_len)
+len(vectorize_layer.get_vocabulary())
+
+
+def create_text_ds(data):
+    texts = [_["text"] for _ in data]
+    text_ds = [vectorize_layer(t) for t in texts]
+    text_ds = tf.data.Dataset.from_tensor_slices(text_ds)
+    return text_ds
+
+
+def path_to_audio(path):
+    # spectrogram using stft
+    audio = tf.io.read_file(path)
+    audio, _ = tf.audio.decode_wav(audio, 1)
+    audio = tf.squeeze(audio, axis=-1)
+    stfts = tf.signal.stft(audio, frame_length=200, frame_step=80, fft_length=256)
+    x = tf.math.pow(tf.abs(stfts), 0.5)
+    # normalisation
+    means = tf.math.reduce_mean(x, 1, keepdims=True)
+    stddevs = tf.math.reduce_std(x, 1, keepdims=True)
+    x = tf.divide(tf.subtract(x, means), stddevs)
+    audio_len = tf.shape(x)[0]
+    # padding to 10 seconds
+    pad_len = 2754
+    paddings = tf.constant([[0, pad_len], [0, 0]])
+    x = tf.pad(x, paddings, "CONSTANT")[:pad_len, :]
+    return x
+
+
+def create_audio_ds(data):
+    flist = [_["audio"] for _ in data]
+    audio_ds = tf.data.Dataset.from_tensor_slices(flist)
+    audio_ds = audio_ds.map(
+        path_to_audio, num_parallel_calls=tf.data.experimental.AUTOTUNE
+    )
+    return audio_ds
+
+
+def create_tf_dataset(data, bs=4):
+    audio_ds = create_audio_ds(data)
+    text_ds = create_text_ds(data)
+    ds = tf.data.Dataset.zip((audio_ds, text_ds))
+    ds = ds.map(lambda x, y: {"src": x, "trg": y})
+    ds = ds.batch(bs)
+    ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
+    return ds
+
+
+"""
+Check contents of one batch
+"""
+
+
+split = int(len(data) * 0.99)
+train_data = data[:split]
+test_data = data[split:]
+
+ds = create_tf_dataset(train_data, bs=64)
+val_ds = create_tf_dataset(test_data, bs=4)
+
+for i in ds.take(1):
+    print(i["src"].shape)
+    print(i["trg"])
+
+
+"""
+## Callbacks to display predictions and to change learning rate
+"""
+
+
+class DisplayOutputs(keras.callbacks.Callback):
+    def __init__(
+        self, batch, idx_to_token, trg_start_token_idx=27, trg_end_token_idx=28
+    ):
+        """ Displays a batch of outputs after every epoch 
+        Arguments:
+        - batch: test batch containing the keys "src" and "trg"
+        - idx_to_token: a List containing the vocabulary tokens corresponding to their indices
+        - trg_start_token_idx: start token index in the target vocabulary
+        - trg_end_token_idx: end token index in the target vocabulary
+        """
```

Comment:
Style guide https://google.github.io/styleguide/pyguide.html


```suggestion
        """ Displays a batch of outputs after every epoch.
        
        Args:
          batch: A test batch containing the keys "src" and "trg".
          idx_to_token: A list containing the vocabulary tokens corresponding to their indices.
          trg_start_token_idx: A start token index in the target vocabulary.
          trg_end_token_idx: An end token index in the target vocabulary.
        """
```